### ISSUE TITLE
[HOLD for payment 2023-02-07] [HOLD #11463] [$2000] [Bug] mWeb/iOS - Request money/Split bill - confirmation button is overlapped by the keyboard

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#10670]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Sign in with any account
2. If needed, start a chat with someone other than concierge by clicking the green plus, New chat, typing in a valid email, and selecting that person.
3. Click the plus at the bottom left, Send money
4. Enter an amount and hit next

## Expected Result:
The pay button appears above the keyboard

## Actual Result:
Final confirmation button is covered by the keyboard. You have to hide keyboard to see the button

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Mobile Web (Safari iOS)

**Version Number:** 1.1.93.4

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/187305737-68ee7a22-6b04-4bf7-99a1-6ba8926beb42.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Beamanator** [comment]
Looks like something a contributor can work on!

**Comment by JmillsExpensify** [comment]
Yes, though I thought that we decided that we're going to treat all keyboard issues centrally at the moment? I'll ask in Slack for confirmation.

**Comment by Beamanator** [comment]
Ohhhhh I think you're right

**Comment by JmillsExpensify** [comment]
Cool, I'll put a hold on this. I also posted in open-source to make sure this one is on people's radars.

**Comment by Beamanator** [comment]
Dropping priority a bit since this is on hold :D

**Comment by Beamanator** [comment]
Still on hold

**Comment by JmillsExpensify** [comment]
Still on hold, though we're going to pick up the larger keyboard initiative soon.

**Comment by JmillsExpensify** [comment]
Still on hold per the larger keyboard initiative.

**Comment by Beamanator** [comment]
Still on hold

**Comment by tgolen** [comment]
Still on hold

**Comment by tgolen** [comment]
This is no longer on hold and can have proposals created for it with the new code.

**Comment by JmillsExpensify** [comment]
Nice, we're open for proposals! I'm going to proactively increase this job to $500 to kick us off.

**Comment by JmillsExpensify** [comment]
Still waiting on proposals.

**Comment by JmillsExpensify** [comment]
Still waiting on proposals. I'm increasing the price to $100. Upwork is here: https://www.upwork.com/jobs/~01eb64954ff199be44.

**Comment by JmillsExpensify** [comment]
Still waiting on proposals.

**Comment by sobitneupane** [comment]
Still waiting on proposals.

**Comment by JmillsExpensify** [comment]
On that note, it's been a week, so it's time to double the price of this issue. Now we're at $2,000. Upwork job is here: https://www.upwork.com/jobs/~01eb64954ff199be44.

**Comment by tgolen** [comment]
Still looking for proposals

**Comment by JmillsExpensify** [comment]
Ideally at this point we get Margelo or an internal engineer to take this over. We're still struggling to get proposals.

**Comment by ntdiary** [comment]
Hi, I'm also interested in this issue and am doing a bit of research these days, expect to post a proposal by next Monday if it's still needed. 😄

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/src/styles/cardStyles/index.js b/src/styles/cardStyles/index.js
index 4759700c6..57cd333d0 100644
--- a/src/styles/cardStyles/index.js
+++ b/src/styles/cardStyles/index.js
@@ -7,10 +7,5 @@ import variables from '../variables';
  * @returns {Object}
  */
 export default function getCardStyles(isSmallScreenWidth, screenWidth) {
-    return {
-        position: 'fixed',
-        width: isSmallScreenWidth ? screenWidth : variables.sideBarWidth,
-        right: 0,
-        height: '100%',
-    };
+    return {};
 }
diff --git a/web/index.html b/web/index.html
index ca117f328..ca59ebfba 100644
--- a/web/index.html
+++ b/web/index.html
@@ -15,7 +15,9 @@
     <link rel="stylesheet" type="text/css" href="/css/fonts.css" />
     <link rel="stylesheet" type="text/css" href="/css/pdf.css" />
     <style>
-        html,
+        html {
+            height: 100%;
+        }
         body,
         #root,
         #root > div,
@@ -50,6 +52,24 @@
     </style>
     <meta name="viewport" content="width=device-width, user-scalable=no, initial-scale=1">
     <link rel="shortcut icon" id="favicon" href="/favicon.png">
+    <script>
+        (function() {
+            var lastScrollPosition = {pageXOffset: 0, pageYOffset: 0};
+            const setLastScrollPosition = () => {
+                lastScrollPosition = {pageXOffset: window.pageXOffset, pageYOffset: window.pageYOffset};
+            }
+            document.addEventListener('scroll', setLastScrollPosition);
+
+            const viewport = window.visualViewport;
+            const setHeightToMatchViewport = () => {
+                document.querySelector("html").style.height = `${viewport.height}px`;
+                window.scrollTo(lastScrollPosition.pageXOffset, lastScrollPosition.pageXOffset);
+            }
+            viewport.addEventListener('resize', setHeightToMatchViewport);
+
+            setHeightToMatchViewport();
+        })();
+    </script>
     <% if (htmlWebpackPlugin.options.usePolyfillIO) { %>
         <!-- polyfill.io is only needed on Web to support older browsers. It should not be loaded for desktop -->
         <script src="https://polyfill.io/v3/polyfill.min.js?features=default%2CResizeObserver&flags=gated"></script>
```

### Details
The issue is caused by three sub-issues:
1. The first issue is that Safari does not always update the `window.innerHeight` property. Debugging exactly why is too much time consuming so I had to find a workaround. Luckily we have [VisualViewport](https://developer.mozilla.org/en-US/docs/Web/API/VisualViewport), so instead of relying on Safari whether it will update the `window.innerHeight` property or not, we get the visual viewport height and apply it to our document. This way we can be sure that we won't have any element behind the keyboard.
2. The second issue is that Safari sometimes pushes up the whole content on keyboard open (e.g. Split bill page). It seems that if the text input is near the top it keeps the page as it is but if the text input is near the bottom the page will be pushed away. Fixing this is a bit tricky, in my proposal I simply do reset the scroll position.
3. The third issue is the behaviour of `position: fixed` on Safari may conflict with the keyboard space, on getCardStyles I had to change the position property but it turns out the whole css block there is redundant so I removed it.

### Tests
Tested on composer input, request money and split bill page.

### Results
The results are as expected and the behaviour is identical to mWeb - Chrome

**Comment by rawalyogendra** [comment]
## Problem
I was looking into this issue and I found out that the problem lies here,

https://github.com/Expensify/App/blob/0294e1c130f1bae588abcb0516cd3d221c300d1a/src/components/withWindowDimensions.js#L49-L51

We are listening to `Dimensions` change event here which should be fired by browser when a keyboard event is triggered, but while testing I found out that Safari is not currently doing that.

It's not just our app, I have created a minimum reproducible example using `react-native-web` example:
https://codesandbox.io/s/stupefied-brook-scdqh6?file=/pages/dimensions/index.js
Preview: https://scdqh6.sse.codesandbox.io/dimensions

And the issue occurs there as well.

If you visit the above URL with Safari and try to use the text input, you will notice that the `window.height` does not change when the keyboard is triggered. But if you visit the same URL with Chrome on an iOS device and try to interact with the text input, you will notice that the `window.height` changes.

This issue is also affecting other parts of our application like “Change Password”, “Add Phone Number” pages.

 I believe this #11463 is also related to the same problem.
 
 As we do not believe in workaround, I think this has to be fixed on `react-native-web` upstream or our own fork.
 
 I currently do not have a solution, but I can look into it.

**Comment by JmillsExpensify** [comment]
@tgolen To help clear out the extensive backlog of /App bugs, we're putting the spotlight on all bugs older than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:

 - Decide whether any proposals currently meet our guidelines and can be approved as-is
 - For any that can't, please take this issue internal and treat it as one of your highest priorities
 - If you have any questions, don't hesitate to start a discussion in #bug-zero

Thanks everyone!

**Comment by sobitneupane** [comment]
@s77rt Thanks for your proposal. 

Issues you mentioned are spot on. 
> The first issue is that Safari does not always update the window.innerHeight property.

I don't think issue is only with Safari. I can reproduce the issue on Chrome as well.

I believe both the issue on request money and split bill has the same root cause. So, as far as possible let's try to find the root cause and solve it.

Having said that, I tested workaround you proposed. But found several issues with it. Your proposed change not only effects IOU flow but whole app as a whole. If you poke around a while, you can find some issues. Some of them being:

- Even after keyboard is closed, white screen (something like keyboard placeholder) remains for a while.
- On scrolling to bottom of the page, you are scrolled back to the top.

**Comment by sobitneupane** [comment]
@rawalyogendra Thanks for your proposal.

>We are listening to Dimensions change event here which should be fired by browser when a keyboard event is triggered, but while testing I found out that Safari is not currently doing that.

I agree with you here.

>If you visit the above URL with Safari and try to use the text input, you will notice that the window.height does not change when the keyboard is triggered. But if you visit the same URL with Chrome on an iOS device and try to interact with the text input, you will notice that the window.height changes.

I tested your mock and the issue exists both on safari and chrome but only on IOS. On android, the height does changes. Same is the case with our app. It is working fine in Android. But on IOS, both on safari and chrome the issue exists.

> As we do not believe in workaround, I think this has to be fixed on react-native-web upstream or our own fork.
I currently do not have a solution, but I can look into it.

Spot on. We definitely prefer the solution that fixes the issue upstream.

**Comment by rawalyogendra** [comment]
>It is working fine in Android. But on IOS, both on safari and chrome the issue exists.

Weird @sobitneupane, let me test this again today on iOS(Chrome) and update my findings

**Comment by sobitneupane** [comment]
I also found another issue while testing above proposals. Keyboard is not opening for the confirmation page. Keyboard only opens after pressing the "What's it for?" `TextInput`. I am not sure if it is already logged somewhere.

https://user-images.githubusercontent.com/25876548/202698411-e5f4dc61-0c8e-44a3-a6e1-5da81a583584.mp4

**Comment by s77rt** [comment]
@sobitneupane 
> I also found another issue while testing above proposals. Keyboard is not opening for the confirmation page. Keyboard only opens after pressing the "What's it for?" TextInput. I am not sure if it is already logged somewhere.

This is being tracked on https://github.com/Expensify/App/issues/10414

**Comment by rawalyogendra** [comment]
@tgolen I have opened an issue in `react-native-web`

https://github.com/necolas/react-native-web/issues/2430

Will Look into it this weekend

**Comment by fedirjh** [comment]
> I don't think issue is only with Safari. I can reproduce the issue on Chrome as well.

Safari thinks this is feature. There is a bug report [here](https://bugs.webkit.org/show_bug.cgi?id=141832)
When you open the keyboard, the browser's window is moved up and your content is hidden
![dBuYh](https://user-images.githubusercontent.com/36869046/202868137-df31d26a-0135-4a50-ab63-bcee63945988.gif)
Check this blog post for more details and more examples of weird behaviors: https://blog.opendigerati.com/the-eccentric-ways-of-ios-safari-with-the-keyboard-b5aa3f34228d

here you can see that not only the final confirmation button is covered by the keyboard, but also the top header is hhidden

**Comment by hammad-6304** [comment]
This issue seems identical to the one i linked below. 

I think if we fix this "KeyboardAvoidingView" upstream then the reported current issue will also be resolved by using the upstream fix. 


I am currently working on this issue to fix (Implementation of Given component) upstream to React-native-web. 

I'll update here shorty i am close to it. 

Thanks. 

CC: @JmillsExpensify @tgolen @tgolen @mananjadhav

**Comment by tgolen** [comment]
OK, thank you for the update! I'm going to place this on HOLD until #11463 is done.

**Comment by JmillsExpensify** [comment]
Great catch, thank you!

**Comment by tjferriss** [comment]
Following instructions for the [weekly update chore](https://stackoverflow.com/c/expensify/questions/8208):

@tgolen this is one of the oldest issues in the /App repo. To help us clear out the large backlog of bugs, can you:

- Decide whether any proposals currently meet our guidelines and can be approved as-is
- For any that can't, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #bug-zero

I recognize this issue is on hold, but once it's not on hold do you expect we should take this internal given the issue's age?

**Comment by JmillsExpensify** [comment]
@tjferriss For the weekly chore, this issue is held on a related issue, so I think we need to ensure that we're bumping the oldest UNHELD issues. As for the hold on this issue, there is a PR awaiting approval/merge in the related issue. Hopefully we'll be back in business soon!

**Comment by tgolen** [comment]
The related PR is in review now and should be merged today if all goes well.

**Comment by rawalyogendra** [comment]
@tgolen @JmillsExpensify should I add my proposal to fix this issue now or after the related PR is merged?

**Comment by JmillsExpensify** [comment]
Still waiting on the proposal [discussed here](https://github.com/Expensify/App/issues/10670#issuecomment-1330850102).

**Comment by rawalyogendra** [comment]
@tgolen @JmillsExpensify added my proposal here

https://github.com/Expensify/App/issues/11463#issuecomment-1336473954

**Comment by JmillsExpensify** [comment]
Discussing in the linked issue about forking now that it's Monday. See #11463 for more context.

**Comment by JmillsExpensify** [comment]
Still working through tests/upstream PR.

**Comment by JmillsExpensify** [comment]
Same as above.

**Comment by JmillsExpensify** [comment]
Still on hold, Melvin.

**Comment by JmillsExpensify** [comment]
We're discussing remaining steps to wrap up the linked issue holding this one. We should be aligned in the coming days on next steps.

**Comment by mallenexpensify** [comment]
Removed `Help Wanted` because job is on hold, please add back when it's off hold

**Comment by JmillsExpensify** [comment]
Closing in on addressing the linked issue. Thanks @mallenexpensify on the label.

**Comment by JmillsExpensify** [comment]
Ok great, that's probably a good reminder for us to get the BugZero checklist kicked off. @sobitneupane and @tgolen mind getting the first three kicked off given that we have a regression in play?

**Comment by sobitneupane** [comment]
@JmillsExpensify I think we can close this issue because it is being handled in https://github.com/Expensify/App/issues/11463

**Comment by JmillsExpensify** [comment]
Huh, that's a good catch. @rawalyogendra Do you agree with this? I'll go ahead and close for now since we're also working on #11463 together, but please let us know if you think this was extra work outside the scope of that linked issue. Thanks!
